### ISSUE TITLE
[bug] auto reload does not update references to class and breaks single dispatch

### ISSUE THREAD

**Issue by PhilipVinc** [ipython/ipython#13258]
TLDR: iPython's auto reload redefines classes that are modified. It also correctly updates the `__class__` of existing instances whose class has been modified, however it does not update references to the class itself. This breaks stuff relying on the class, like `@singledispatch`.

Consider the following example file:
```python
➜ cat bug.py
from dataclasses import dataclass

class A:
    a : int = 1

    # uncomment after autoreload
    #def __repr__(self):
    #    return f"newA({self.a})"
```

If I now run a ipython session 

```python
> %load_ext autoreload
> %autoreload 2
> from functools import singledispatch
> import bug
>@singledispatch
>def fun(x):
>    print(f"unknown {type(x)}: {x}")

>@fun.register
>def _a(x: A):
>    print(f"A! {type(x)}: {x}")

> a = bug.A()
> print(a)
<bug.A object at 0x1077d6ac0>
> fun(a)
A! <class 'bug.A'>: <bug.A object at 0x1077d6ac0 >
```

this works fine.
If now I uncomment the 3 lines in the `bug.py` file and save it, you will immediately notice that the class of a is correctly updated (as testified by the new `__repr__`) but the dispatch will not hit the `_a` method while instead falling back to the generic implementation

```python
> print(a)
newA(1)
> fun(a)
unknown <class 'bug.A'>: newA(1)
```

I believe that it should be relatively straightforward to fix this by also checking to stored references to the class itself and updating it
